### PR TITLE
Allow contact to set DNC flag in preferences center

### PR DESCRIPTION
**Pull request by patrykgruszka**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | N
| Automated tests included? | N
| Related user documentation PR URL | https://github.com/mautic/mautic-documentation/pull/88
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | N
| Deprecations? | N

[//]: # ( Required: )
#### Description:
When preference center is activated there is no option for user to unsubscribe from all marketing messages. This PR allows user to set Do Not Contact flag by clicking link in subscription center. This is optional and can be turned off in Mautic configuration.

image

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Activate Preferences center and Do Not Contact option in Mautic settings
3. Send email to user with unsubscribe link
4. Open preferences center (click unsubscribe link)
5. Click Unsubscribe from all marketing messages link
6. Contact should have DNC flag on

### CODE REVIEW

File: app/bundles/CoreBundle/Views/Slots/donotcontact.html.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<?php
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
```
Comment (kuzmany) [inline app/bundles/CoreBundle/Views/Slots/donotcontact.html.php]: Should be 2019

Comment (patrykgruszka) [reply app/bundles/CoreBundle/Views/Slots/donotcontact.html.php]: Right, thanks for update

**Comment by dongilbert** [comment]
This is great - I would like to see a per-channel DNC option (Email, Social, SMS, and Mobile Push), along side the option to DNC on all channels. Is that something you could work up?

**Comment by patrykgruszka** [comment]
> This is great - I would like to see a per-channel DNC option (Email, Social, SMS, and Mobile Push), along side the option to DNC on all channels. Is that something you could work up?

@dongilbert thanks for your opinion. I can't work on per-channel DNC feature right now, but seems like good idea. By the way, is version 2.x development still in progress?

**Comment by dongilbert** [comment]
2.x is in more or less maintenance development. We're going to try to bring feature code from 2.x forward to 3.x, however.

3.x development is in full swing and we're looking at getting an alpha out to the community soon.

**Comment by dongilbert** [comment]
As for the PR, that's OK that you don't have time - as a fellow open source contributor, I get it :D Thanks for your work so far!

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 minor versions (which means no more features in M2) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.1.0 candidate.**

## How to do?
1. Check if your feature or enhancement is still missing and relevant in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please consider closing your PR if you are absolutely sure that the feature has made it into Mautic 3 already

**Please report results by commenting on your PR to make us administration easier.**
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by patrykgruszka** [comment]
Hi @npracht, rebased and tested on 3.x. Works fine after few adjustments.

**Comment by npracht** [comment]
Thanks a lot @patrykgruszka !

**Comment by RCheesley** [comment]
Bumping this to 3.2 - it is going to need a rebase to get the code coverage test report (but best to wait for the 3.1 merges to be done first!) and two good tests on 3.x before we can merge.

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by patrykgruszka** [comment]
This PR is still relevant to the latest version of Mautic

**Comment by RCheesley** [comment]
Thanks @patrykgruszka - are you able to address the test coverage so that we can get the coverage to not change or improve?

**Comment by patrykgruszka** [comment]
@RCheesley I'll check it out

**Comment by RCheesley** [comment]
@patrykgruszka we are hoping to get this into the 4.2 release, so if you can take a look at the test coverage it would be much appreciated!

**Comment by kuzmany** [comment]
@patrykgruszka I am testing it.

What is different between 

`Text for the {unsubscribe_text} token`
`Text for the {do_not_contact_text} token`

?

For me both do same.

image

**Comment by patrykgruszka** [comment]
@kuzmany it makes sense when you use contact preference settings. You can either use link to the preference settings page or immediately unsubscribe from all channels using link from {do_not_contact_text}.

**Comment by patrykgruszka** [comment]
This text is also displayed in the bottom section of the preference settings to allow contact to unsubscribe from all channels.

Here is a screenshot from preference settings page with this text and visual modifications:
obraz

**Comment by kuzmany** [comment]
@patrykgruszka  Thank you for clarification

Yes, I understood this Unsubscribe from all marketing messages on preferences page.

What are other features what you mention?

> You can either use link to the preference settings page or immediately unsubscribe from all channels using link from {do_not_contact_text}.

Should I just test `{dnc_url}` and `{do_not_contact_text}`?

If I did test it unsubscribe me just from emails, not SMS... Do it unsubscribe from all channels?

**Comment by kuzmany** [comment]
As I digging code I see getUnsubscribeMessage method called in you  doNotContactAction 

https://github.com/mautic/mautic/pull/8242/files#diff-19429c7a4f887b40880788b54e72b2e4e996b6917b86c24ca098b1cd6d3a0ea8R356

It does same what `{unsubscribe_url}` and `{unsubscribe_text}` Why why do not allow use these tokens on preference page, why we create token with same behaviour? 

Also Unsubcribe from all marketing channels should really unsubscribe from all channels, not just emails. Then probably It will make more sense. What do you think?

Do I miss something?

**Comment by patrykgruszka** [comment]
@kuzmany the `getUnsubscribeMessage` method is unreachable in `unsubscribeAction` when preference center is activated. Actually the names are unfortunate, because `unsubscribeAction` page is becomes preferences settings when it's activated and unsubscribing from all emails is no longer possible.

When contact preferences option is active, then:
* `{unsubscribe_url}` and `{unsubscribe_text}` links to contact preferences settings page
* `{dnc_url}` and `{do_not_contact_text}` allows user to immediately unsubscribe from all channels

So you can also use it this way in your email to allow contact to immediately unsubscribe from all:
> Click `{unsubscribe_url}` to manage your preferences or unsubscribe from all messages `{dnc_url}`.

When the contact preferences settings is turned off then this tokens have the same behaviour, as you said. And I have no idea how to implement it better without BC. 

> Also Unsubcribe from all marketing channels should really unsubscribe from all channels, not just emails. Then probably It will make more sense. What do you think?

You are right, unsubscribing from all channels would be more useful. I think we can change this behaviour.

I see some inconsistency in the tag names `{dnc_url}` and `{do_not_contact_text}`, this should be fixed as well. 

If you have any other ideas how to make this changes more clear and useful please let me know.

**Comment by kuzmany** [comment]
@patrykgruszka thank now it's clear. We would like it use in production with much more configuration of DNC improvements. Could I finish this? It looks like need just some minor touch

**Comment by patrykgruszka** [comment]
@kuzmany sure, go ahead.

**Comment by kuzmany** [comment]
I cherry pick and finish it with https://github.com/mautic/mautic/pull/10579 
Probably we can close this PR

- I've added token Unsubscribe to page builder
- I fixed unsubscribe all to DNC all available channels (email, sms) and return to landing page with success message

**Pull request closed without merging**